File: simplified-books-core/src/main/java/org/nypl/simplified/books/core/BookStatusRevokeFailed.java
package org.nypl.simplified.books.core;

import com.io7m.jfunctional.OptionType;
import com.io7m.jnull.NullCheck;

/**
 * The given book could not be revoked (hold cancelled, loan returned, etc).
 */

public final class BookStatusRevokeFailed implements BookStatusType
{
  private final OptionType<Throwable> error;
  private final BookID                id;

  /**
   * Construct a status value.
   *
   * @param in_id            The book ID
   * @param x                The exception raised, if any
   */

  public BookStatusRevokeFailed(
    final BookID in_id,
    final OptionType<Throwable> x)
  {
    this.id = NullCheck.notNull(in_id);
    this.error = NullCheck.notNull(x);
  }

  @Override public BookID getID()
  {
    return this.id;
  }

  /**
   * @return error
   */
  public OptionType<Throwable> getError()
  {
    return this.error;
  }

  @Override public <A, E extends Exception> A matchBookStatus(
    final BookStatusMatcherType<A, E> m)
    throws E
  {
    return m.onBookStatusRevokeFailed(this);
  }

  @Override public String toString()
  {
    final StringBuilder b = new StringBuilder(128);
    b.append("[BookStatusRevokeFailed ");
    b.append(this.id);
    b.append(" ");
    b.append(this.error);
    b.append("]");
    return NullCheck.notNull(b.toString());
  }

  @Override public BookStatusPriorityOrdering getPriority()
  {
    return BookStatusPriorityOrdering.BOOK_STATUS_REVOKE_FAILED;
  }
}

File: simplified-books-registry-api/src/main/java/org/nypl/simplified/books/book_registry/BookStatusRevokeFailed.kt
package org.nypl.simplified.books.book_registry

import org.nypl.simplified.books.api.BookID
import org.nypl.simplified.presentableerror.api.PresentableErrorType
import org.nypl.simplified.presentableerror.api.Presentables
import org.nypl.simplified.taskrecorder.api.TaskResult

/**
 * The given book could not be revoked (hold cancelled, loan returned, etc).
 */

data class BookStatusRevokeFailed(

  /**
   * The book ID
   */

  val id: BookID,

  /**
   * The list of steps that lead to the failure.
   */

  val result: TaskResult.Failure<BookStatusRevokeErrorDetails, Unit>)
  : PresentableErrorType, BookStatusType {

  override val attributes: Map<String, String>
    get() = Presentables.collectAttributes(this.result.errors())

  override val message: String
    get() = this.result.steps.last().resolution.message

  override fun getID(): BookID =
    this.id

  override fun <A, E : Exception> matchBookStatus(m: BookStatusMatcherType<A, E>): A =
    m.onBookStatusRevokeFailed(this)

  override fun toString(): String {
    val b = StringBuilder(128)
    b.append("[BookStatusRevokeFailed ")
    b.append(this.id)
    b.append("]")
    return b.toString()
  }

  override fun getPriority(): BookStatusPriorityOrdering =
    BookStatusPriorityOrdering.BOOK_STATUS_REVOKE_FAILED
}
